src/components/tl/sharp/tl_sharp_context.c

```diff
@@ -0,0 +1,323 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include <inttypes.h>
+#include "tl_sharp.h"
+
+static ucc_mpool_ops_t ucc_tl_sharp_req_mpool_ops = {
+    .chunk_alloc   = ucc_mpool_hugetlb_malloc,
+    .chunk_release = ucc_mpool_hugetlb_free,
+    .obj_init      = NULL,
+    .obj_cleanup   = NULL
+};
+
+static int ucc_tl_sharp_oob_barrier(void *arg)
+{
+    ucc_tl_sharp_oob_ctx_t *oob_ctx = (ucc_tl_sharp_oob_ctx_t *)arg;
+    ucc_tl_sharp_context_t *ctx     = oob_ctx->ctx;
+    ucc_oob_coll_t *oob_coll;
+    ucc_status_t status;
+    char sbuf;
+    void *req;
+    char *rbuf;
+
+    oob_coll = (oob_ctx->ctx_oob != NULL) ? oob_ctx->ctx_oob: oob_ctx->team_oob;
+    rbuf     = ucc_malloc(sizeof(char) * oob_coll->n_oob_eps, "tmp_barrier");
+    if (!rbuf) {
+        tl_error(ctx->super.super.lib,
+                 "failed to allocate %ld bytes for tmp barrier array",
+                 sizeof(char) * oob_coll->n_oob_eps);
+        return -1;
+    }
+
+    status = oob_coll->allgather(&sbuf, rbuf, sizeof(char),
+                                oob_coll->coll_info, &req);
+    if (UCC_OK == status) {
+        ucc_assert(req);
+        while (UCC_OK != (status = oob_coll->req_test(req))) {
+            if (status < 0) {
+                tl_error(ctx->super.super.lib, "failed to test oob req");
+                break;
+            }
+        }
+        oob_coll->req_free(req);
```

Comment:
Generic comment: this is a blocking call, which makes sharp_Ctx/comm creation blocking as well. We don't progress runtime here. Can it lead to potential issues?